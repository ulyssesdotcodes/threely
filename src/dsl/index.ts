// Main DSL exports - combines all chain types and utilities
import { MockObject3D, mockUtils, mockPresets } from "../three/MockObject3D";

// Object3D chain exports - pure functions (no Node dependencies)
export {
  // Scene management
  setScene,
  getObjectRegistry,
  clearAll,

  // Geometry creation
  sphere,
  box,
  cylinder,
  material,
  mesh,

  // Transforms
  translateX,
  translateY,
  translateZ,
  rotateX,
  rotateY,
  rotateZ,
  applyMock,

  // Rendering
  render,

  // Chain object
  chainObj3d,
} from "./pure-object3d-functions";

// Math chain exports - pure functions (no Node dependencies)
export {
  // Basic operations
  mult,
  add,
  sub,
  div,
  frame,

  // Math functions
  mathAbs,
  mathAcos,
  mathAcosh,
  mathAsin,
  mathAsinh,
  mathAtan,
  mathAtan2,
  mathAtanh,
  mathCbrt,
  mathCeil,
  mathClz32,
  mathCos,
  mathCosh,
  mathExp,
  mathExpm1,
  mathFloor,
  mathFround,
  mathHypot,
  mathImul,
  mathLog,
  mathLog10,
  mathLog1p,
  mathLog2,
  mathMax,
  mathMin,
  mathPow,
  mathRandom,
  mathRound,
  mathSign,
  mathSin,
  mathSinh,
  mathSqrt,
  mathTan,
  mathTanh,
  mathTrunc,

  // Chain object
  chainMath,
} from "./pure-math-functions";

// Parser exports
export {
  parseDSLWithLezer,
  executeDSL,
  logToPanel,
  dslContext,
} from "./parser";

// Converter exports
export { convertASTToNodysseus } from "./direct-ast-to-nodysseus-converter";

// Re-export utilities
export { MockObject3D, mockUtils, mockPresets };

// Individual function imports removed - now only using pure functions without Node dependencies
