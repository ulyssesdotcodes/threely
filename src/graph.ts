// graph.ts - Implementation of a node-based graph system with TypeScript

  /**
   * Interface representing a Node in the graph.
   *
   * @template T - The type of data associated with this node
   */
  export interface Node<T = any> {
    id: string;
    data?: T | ((...args: any[]) => void);
    inputs: Map<string, Node<T>>;
    outputs: Map<string, Node<T>>;
  }

  /**
   * Interface for a FunctionNode that has a function as its data
   */
  export interface FunctionNode<T = any> extends Node<T> {
    data: (...args: any[]) => void;
  }

  /**
   * Type guard to check if a node is a FunctionNode
   */
  export function isFunctionNode<T>(node: Node<T>): node is FunctionNode<T> {
    return typeof node.data === 'function' && node.data.length >= 0;
  }

  /**
   * Generic Graph class that manages nodes and their connections.
   *
   * @template T - The type of data associated with the graph's nodes
   */
  export class Graph<T = any> {
    private nodes: Map<string, Node<T>>;
    // Make currentNode accessible from outside the class but read-only
    private _currentNode?: Node<T>;
    get currentNode(): Node<T> | undefined {
      return this._currentNode;
    }
    set currentNode(node: Node<T> | undefined) {
      this._currentNode = node;
    }

    constructor() {
      this.nodes = new Map();
      this.currentNode = undefined;
    }

    /**
     * Create a new node with unique ID
     * @param data - Optional data to associate with the node
     * @returns The created node
     */
    createNode(data?: T | (() => void)): Node<T> {
      const id = this.generateUniqueId();
      const node: Node<T> = { id, data, inputs: new Map(), outputs: new Map() };
      this.nodes.set(id, node);
      return node;
    }

    /**
     * Connect two nodes via named inputs
     * @param source - The source node to connect from
     * @param target - The target node to connect to
     * @param inputName - The name of the input on the target node
     */
    connect(source: Node<T>, target: Node<T>, inputName: string): void {
      if (!target.inputs.has(inputName)) {
        target.inputs.set(inputName, source);
      } else {
        console.warn(`Input "${inputName}" already has a connection`);
      }

      source.outputs.set(inputName, target);
    }

    /**
     * Generate a UUID for nodes (RFC4122 version 4 compliant)
     * @returns A unique string ID in UUID format
     */
    private generateUniqueId(): string {
      // RFC4122, version 4 compliant UUID generator
      let d = new Date().getTime();
      if (typeof performance !== 'undefined' && typeof performance.now === 'function'){
        d += performance.now(); //use high-precision timer if available
      }
      return 'xxxxxxxx-xxxx-4xxx-yxxx-xxxxxxxxxxxx'.replace(/[xy]/g, function(c) {
        const r = (d + Math.random()*16)%16 | 0;
        d = Math.floor(d/16);
        return (c=='x' ? r : (r&0x3|0x8)).toString(16);
      });
    }

    /**
     * Get a node by its ID
     * @param id - The ID of the node to retrieve
     * @returns The node with the specified ID, or undefined if not found
     */
    getNode(id: string): Node<T> | undefined {
      return this.nodes.get(id);
    }

    /**
     * Get all nodes in the graph (for demonstration purposes)
     * @returns An array of all nodes
     */
    getAllNodes(): Node<T>[] {
      return Array.from(this.nodes.values());
    }

    /**
     * Helper function to recursively run input nodes and collect their results
     * @param node - The node whose inputs should be executed
     * @returns An array of results from executing the input nodes' data functions
     */
    private executeInputs(node: Node<T>): any[] {
      const results = [];
      for (const [inputName, inputNode] of node.inputs) {
        if (isFunctionNode(inputNode)) {
          // Recursively run inputs and get their results
          const inputResults = this.executeInputs(inputNode);
          // Execute the input node's data function with its input results
          const result = inputNode.data ? inputNode.data(...inputResults) : undefined;
          results.push(result);
        } else {
          results.push(undefined);
        }
      }
      return results;
    }

    /**
     * Run the graph starting from a root node
     * @param root - The root node to start execution from
     * @returns The result of executing the root node's data function
     */
    run(root: Node<T>): any {
      // Get input values by recursively running all inputs
      const args = this.executeInputs(root);

      // Execute the root node's data function if it exists and is a function
      if (isFunctionNode(root) && typeof root.data === 'function') {
        return root.data(...args);
      }

      return undefined;
    }
  }

  /**
   * Function to wrap regular functions into graph nodes
   *
   * @param fn - The function to wrap
   * @param graph - The graph instance to use for node creation
   * @returns A wrapped function that creates a node when called
   */
  export function createGraphNodeWrapper<T>(
    fn: (...args: any[]) => void,
    graph?: Graph<T>
  ): (...args: any[]) => void {
    // Use the provided graph or default to globalGraph if none is provided
    const targetGraph = graph || (global as any).globalGraph;

    // Create a proxy handler for the function
    const handler: ProxyHandler<any> = {
      apply(target, thisArg, argumentsList) {
        // Save the previous current node
        const prevNode = targetGraph.currentNode;

        // Create a node for this function call
        const node = targetGraph.createNode(() => target.apply(thisArg, argumentsList));

        // If there's a previous node in the chain, connect it to this one
        if (prevNode) {
          targetGraph.connect(prevNode, node, 'next');
        }

        // Set this as the current node for the next call
        targetGraph.currentNode = node;

        // Execute the function and return its result
        const result = target.apply(thisArg, argumentsList);

        // Reset to the previous current node after execution
        targetGraph.currentNode = prevNode;

        return result;
      }
    };

    // Create a proxy that wraps the original function
    const wrappedFn = new Proxy(fn, handler);

    return wrappedFn;
  }

  /**
   * Example usage of the decorator function
   */
  class Example {
    exampleMethod = createGraphNodeWrapper(function() {
      console.log('This is an example method');
    }).bind(this);

    anotherMethod = createGraphNodeWrapper(function() {
      console.log('This is another method in the chain');
    }).bind(this);
  }

  // Export for testing
  export { Graph as DefaultGraph };