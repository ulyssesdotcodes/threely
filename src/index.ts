import { initBackgroundScene } from './three-render';
import { startEditor, setupEditorUI } from './editor';
import { initErrorDisplay } from './error-display';

document.addEventListener('DOMContentLoaded', () => {
  // Initialize error display system first
  initErrorDisplay();
  
  initBackgroundScene();
  startEditor();
  setupEditorUI();
});