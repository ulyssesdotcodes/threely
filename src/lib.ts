// Main library exports for npm publishing
export { parseDSL, executeDSL, dslContext, setScene, getObjectRegistry, clearAll } from './dsl';
// Deprecated: Graph system replaced by direct AST to Nodysseus conversion
// export { Graph, Node, createNode, map, apply } from './graph';
// export type { Node as ThreelyNode } from './graph';
export * as THREE from 'three';