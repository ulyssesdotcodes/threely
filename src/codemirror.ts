import { EditorState, Prec, Compartment } from "@codemirror/state";
import { EditorView, keymap } from "@codemirror/view";
import { basicSetup } from "codemirror";
import { javascript } from "@codemirror/lang-javascript";
import { vim } from "@replit/codemirror-vim";
import { getTextBlockAtPosition } from "./text_utils";
import { executeDSL } from "./dsl";
import {
  uuidRangeSetField,
  uuidRangeSetPlugin,
  generateUUIDTags,
  setUUIDRangeSet,
  getUUIDFromState,
  UUIDTag,
} from "./uuid-tagging";
import { State } from "./nodysseus/node-types";

export { EditorState, EditorView, keymap, basicSetup, javascript };

// Vim mode state management
const VIM_MODE_KEY = "three-tree-vim-mode";
const vimCompartment = new Compartment();
let currentEditorView: EditorView | null = null;

export function getVimModeEnabled(): boolean {
  try {
    return localStorage.getItem(VIM_MODE_KEY) === "true";
  } catch {
    return false;
  }
}

export function setVimModeEnabled(enabled: boolean): void {
  try {
    localStorage.setItem(VIM_MODE_KEY, enabled.toString());
    if (currentEditorView) {
      currentEditorView.dispatch({
        effects: vimCompartment.reconfigure(enabled ? vim() : []),
      });
    }
  } catch {
    // Ignore localStorage errors
  }
}

export function getCurrentEditorView(): EditorView | null {
  return currentEditorView;
}

// Custom command for Ctrl-Enter
const handleCtrlEnter = (view: EditorView): boolean => {
  const blockInfo = getBlockAtCursor(view);

  if (blockInfo && blockInfo.block) {
    const code = blockInfo.block.trim();

    try {
      // const ranges: { start: number; end: number; uuid: UUIDTag }[] = [];
      // .between(
      //   blockInfo.start,
      //   blockInfo.end,
      //   (start, end, uuid) => (
      //     ranges.push({
      //       start: start - blockInfo.start,
      //       end: end - blockInfo.start,
      //       uuid,
      //     }),
      //     undefined
      //   ),
      // );
      const result = executeDSL(
        code,
        view.state.field(uuidRangeSetField),
        undefined,
        blockInfo.start,
      );
      if (result) {
      } else {
      }
    } catch (error) {
      console.error("Error executing DSL code:", error);
    }
  }

  // Prevent the default new line behavior by returning true
  return true;
};

// Function to get block at cursor position
export const getBlockAtCursor = (
  view: EditorView,
): { block: string; start: number; end: number } | null => {
  const text = view.state.doc.toString();
  const cursorPos = view.state.selection.ranges[0].from;

  return getTextBlockAtPosition(text, cursorPos);
};

export const defaultContent = `mesh(sphere(), material()).translateX(1).rotateY(45).render("mySphere")

mesh(sphere(), material()).translateX(frame().mult(0.02).floor()).rotateY(45).render("mySphere")

// Try pressing Ctrl+Enter on the line above!
// This will create a sphere mesh named "mySphere", translate it, rotate it, and add it to the scene
// Running it again will update the existing object instead of creating a new one!
// Note: sphere(), box(), and cylinder() now create mock geometries that are applied during render!

mesh(box(2, 1, 1), material({color: 0xff0000})).translateX(-3).render("redBox")

mesh(cylinder(), material({color: 0x0000ff, wireframe: true})).translateX(3).render("blueCylinder")

// Animated sphere using frame counter:
mesh(sphere(0.5), material({color: 0x00ff00})).translateX(Math.sin(frame() * 0.01) * 2).render("animatedSphere")

// Using mock objects for quick property setting:
applyMock(mesh(sphere(), material({color: 0xff0000})), mockUtils.position(2, 1, 0)).render("mockSphere")

// Using presets for common configurations:
applyMock(mesh(box(), material()), mockPresets.elevated(3)).render("elevatedBox")

// Custom geometry parameters with mock system:
mesh(sphere(1.5, 16, 8), material({color: 0xff00ff})).render("customSphere")

// Custom box with specific dimensions:
mesh(box(3, 0.5, 0.5), material({color: 0x00ffff})).translateY(-2).render("customBox")

// Custom cylinder with different top/bottom radius:
mesh(cylinder(0.5, 1, 2), material({color: 0xffff00})).translateZ(-3).render("customCylinder")

// Try modifying the values and re-running to see objects update:
// mesh(sphere(0.5), material({color: 0xffffff})).translateY(2).rotateX(90).render("mySphere")
// mesh(box(1, 3, 1), material({color: 0x00ff00})).translateX(-2).render("redBox")

// Use clearAll() to remove all objects:
// clearAll()
`;

export function createEditorState(
  content: string = defaultContent,
): EditorState {
  const vimModeEnabled = getVimModeEnabled();

  // Generate new tags for the content
  const { rangeSet } = generateUUIDTags(content);

  const state = EditorState.create({
    doc: content,
    extensions: [
      vimCompartment.of(vimModeEnabled ? vim() : []),
      basicSetup,
      javascript(),
      Prec.highest(keymap.of([{ key: "Ctrl-Enter", run: handleCtrlEnter }])),
      uuidRangeSetField, // Add UUID RangeSet field
      uuidRangeSetPlugin, // Add UUID RangeSet plugin for automatic updates
    ],
  });

  // Apply the initial UUID RangeSet
  return state.update({
    effects: setUUIDRangeSet.of(rangeSet),
  }).state;
}

export function setCurrentEditorView(view: EditorView): void {
  currentEditorView = view;
}
