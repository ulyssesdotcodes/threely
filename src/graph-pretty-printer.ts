// graph-pretty-printer.ts - Pretty printing functionality for graph visualization
import { Node } from './graph';

/**
 * Options for pretty printing the graph
 */
export type PrettyPrintOptions = {
  /** Show node IDs in the output */
  readonly showIds?: boolean;
  /** Maximum depth to traverse (prevents infinite loops) */
  readonly maxDepth?: number;
  /** Custom node label function */
  readonly nodeLabel?: (node: Node<any>) => string;
  /** Include dependency count */
  readonly showDependencyCount?: boolean;
}

/**
 * Pretty printer for graph visualization
 */
export class GraphPrettyPrinter {
  private visited = new Set<string>();
  private options: Required<PrettyPrintOptions>;

  constructor(options: PrettyPrintOptions = {}) {
    this.options = {
      showIds: options.showIds ?? true,
      maxDepth: options.maxDepth ?? 10,
      nodeLabel: options.nodeLabel ?? this.defaultNodeLabel.bind(this),
      showDependencyCount: options.showDependencyCount ?? true
    };
  }

  /**
   * Default node label generator
   */
  private defaultNodeLabel(node: Node<any>): string {
    const computeStr = node.compute.toString();
    
    // Try to extract meaningful info from the compute function
    if (computeStr.includes('SphereGeometry')) return 'sphere';
    if (computeStr.includes('BoxGeometry')) return 'box';
    if (computeStr.includes('CylinderGeometry')) return 'cylinder';
    if (computeStr.includes('MeshBasicMaterial')) return 'material';
    if (computeStr.includes('THREE.Mesh') || computeStr.includes('new THREE.Mesh')) return 'mesh';
    if (computeStr.includes('translateXObj')) return 'translateX';
    if (computeStr.includes('translateYObj')) return 'translateY';
    if (computeStr.includes('translateZObj')) return 'translateZ';
    if (computeStr.includes('rotateXObj')) return 'rotateX';
    if (computeStr.includes('rotateYObj')) return 'rotateY';
    if (computeStr.includes('rotateZObj')) return 'rotateZ';
    if (computeStr.includes('currentScene.add') || computeStr.includes('objectRegistry')) return 'render';
    if (computeStr.includes('Graph.run')) return 'map';
    
    return 'node';
  }

  /**
   * Pretty print a single node with its dependencies
   */
  print(node: Node<any>, depth: number = 0): string {
    const indent = '  '.repeat(depth);
    const nodeId = this.options.showIds ? ` (${node.id})` : '';
    const depCount = this.options.showDependencyCount ? ` [deps: ${node.dependencies.length}]` : '';
    const label = this.options.nodeLabel(node);
    
    let result = `${indent}${label}${nodeId}${depCount}\n`;

    // Prevent infinite recursion
    if (depth >= this.options.maxDepth) {
      return result + `${indent}  ...(max depth reached)\n`;
    }

    // Prevent cycles
    if (this.visited.has(node.id)) {
      return result + `${indent}  ...(already visited)\n`;
    }

    this.visited.add(node.id);

    // Print dependencies
    for (let i = 0; i < node.dependencies.length; i++) {
      const dep = node.dependencies[i];
      const isLast = i === node.dependencies.length - 1;
      const connector = isLast ? '└─ ' : '├─ ';
      
      result += `${indent}${connector}`;
      result += this.print(dep, depth + 1).slice(indent.length + connector.length);
    }

    return result;
  }

  /**
   * Print multiple nodes as a forest
   */
  printForest(nodes: Node<any>[]): string {
    this.visited.clear();
    let result = '';
    
    for (let i = 0; i < nodes.length; i++) {
      if (i > 0) result += '\n';
      result += `Root ${i + 1}:\n`;
      result += this.print(nodes[i]);
    }
    
    return result;
  }

  /**
   * Create a compact one-line representation
   */
  compact(node: Node<any>): string {
    const label = this.options.nodeLabel(node);
    const deps = node.dependencies.map(dep => this.compact(dep)).join(', ');
    
    if (deps) {
      return `${label}(${deps})`;
    }
    return label;
  }
}