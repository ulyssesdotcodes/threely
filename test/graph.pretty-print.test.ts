import { Graph } from '../src/graph';
import { GraphPrettyPrinter } from '../src/graph-pretty-printer';
import { sphere, box, material, mesh, translateX, translateY, rotateX, rotateY, render } from '../src/dsl';

describe('Graph Pretty Printing', () => {
  it('should pretty print a complex graph with approximately 10 nodes', () => {
    // Create a complex scene with multiple objects
    // This will create approximately 10 nodes total
    
    // First object: sphere with transformations (5 nodes)
    const sphereGeom = sphere(1, 16, 12);           // node 1
    const redMaterial = material({ color: 0xff0000 }); // node 2
    const sphereMesh = mesh(sphereGeom, redMaterial);   // node 3
    const movedSphere = translateX(sphereMesh, 2);      // node 4
    const rotatedSphere = rotateY(movedSphere, Math.PI / 4); // node 5
    
    // Second object: box with different transformations (5 nodes)
    const boxGeom = box(2, 2, 2);                   // node 6
    const blueMaterial = material({ color: 0x0000ff }); // node 7
    const boxMesh = mesh(boxGeom, blueMaterial);        // node 8
    const liftedBox = translateY(boxMesh, 3);           // node 9
    const finalScene = render(liftedBox, "testBox");   // node 10
    
    // Test default pretty print
    const output = Graph.prettyPrint(finalScene);
    
    expect(output).toMatchSnapshot();
  });

  it('should pretty print with custom options', () => {
    // Create a smaller graph for custom options test
    const sphereGeom = sphere(2);
    const mat = material({ color: 0x00ff00 });
    const meshNode = mesh(sphereGeom, mat);
    const transformed = rotateX(meshNode, Math.PI / 2);
    const rendered = render(transformed, "customTest");
    
    const output = Graph.prettyPrint(rendered, {
      showIds: false,
      showDependencyCount: false,
      nodeLabel: (node) => {
        const computeStr = node.compute.toString();
        if (computeStr.includes('SphereGeometry')) return 'SPHERE_GEOMETRY';
        if (computeStr.includes('MeshBasicMaterial')) return 'BASIC_MATERIAL';
        if (computeStr.includes('THREE.Mesh')) return 'MESH_OBJECT';
        if (computeStr.includes('rotateXObj')) return 'ROTATE_X_TRANSFORM';
        if (computeStr.includes('currentScene.add')) return 'SCENE_RENDER';
        return 'UNKNOWN_NODE';
      }
    });
    
    expect(output).toMatchSnapshot();
  });

  it('should create compact representation', () => {
    // Create a multi-branch graph
    const geom1 = sphere(1);
    const geom2 = box(1, 1, 1);
    const mat1 = material({ color: 0xff0000 });
    const mat2 = material({ color: 0x00ff00 });
    
    const mesh1 = mesh(geom1, mat1);
    const mesh2 = mesh(geom2, mat2);
    
    const transformed1 = translateX(mesh1, -2);
    const transformed2 = translateX(mesh2, 2);
    
    const rendered1 = render(transformed1, "left");
    const rendered2 = render(transformed2, "right");
    
    const compact1 = Graph.compact(rendered1);
    const compact2 = Graph.compact(rendered2);
    
    expect(compact1).toMatchSnapshot();
    expect(compact2).toMatchSnapshot();
  });

  it('should handle forest printing with multiple root nodes', () => {
    // Create multiple independent graphs
    const tree1 = render(
      rotateY(
        translateX(
          mesh(sphere(1), material({ color: 0xff0000 })),
          1
        ),
        Math.PI / 6
      ),
      "tree1"
    );
    
    const tree2 = render(
      translateY(
        mesh(box(1, 2, 1), material({ color: 0x00ff00 })),
        -1
      ),
      "tree2"
    );
    
    const tree3 = mesh(sphere(0.5), material({ wireframe: true }));
    
    const printer = new GraphPrettyPrinter();
    const forestOutput = printer.printForest([tree1, tree2, tree3]);
    
    expect(forestOutput).toMatchSnapshot();
  });

  it('should show dependency chain execution', () => {
    // Create a linear dependency chain to show execution order
    const baseGeometry = sphere(1, 8, 6);
    const baseMaterial = material({ color: 0xffffff });
    const baseMesh = mesh(baseGeometry, baseMaterial);
    
    // Chain of transformations
    const step1 = translateX(baseMesh, 1);
    const step2 = translateY(step1, 2);
    const step3 = rotateX(step2, Math.PI / 4);
    const step4 = rotateY(step3, Math.PI / 3);
    const final = render(step4, "chainTest");
    
    const output = Graph.prettyPrint(final, {
      showIds: true,
      showDependencyCount: true
    });
    
    expect(output).toMatchSnapshot();
  });
});