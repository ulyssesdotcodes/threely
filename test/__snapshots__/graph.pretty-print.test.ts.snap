// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Graph Pretty Printing should create compact representation 1`] = `"mesh(function(function(constant, material)))"`;

exports[`Graph Pretty Printing should create compact representation 2`] = `"mesh(function(function(constant, material)))"`;

exports[`Graph Pretty Printing should handle forest printing with multiple root nodes 1`] = `
"Root 1:
mesh (node-31) [deps: 1]
└─ unction (node-30) [deps: 1]
  └─ unction (node-29) [deps: 1]
    └─ unction (node-28) [deps: 2]
      ├─ onstant (node-26) [deps: 0]
      └─ aterial (node-27) [deps: 0]

Root 2:
mesh (node-36) [deps: 1]
└─ unction (node-35) [deps: 1]
  └─ unction (node-34) [deps: 2]
    ├─ onstant (node-32) [deps: 0]
    └─ aterial (node-33) [deps: 0]

Root 3:
function (node-39) [deps: 2]
├─ onstant (node-37) [deps: 0]
└─ aterial (node-38) [deps: 0]
"
`;

exports[`Graph Pretty Printing should pretty print a complex graph with approximately 10 nodes 1`] = `
"mesh (node-10) [deps: 1]
└─ unction (node-9) [deps: 1]
  └─ unction (node-8) [deps: 2]
    ├─ onstant (node-6) [deps: 0]
    └─ aterial (node-7) [deps: 0]
"
`;

exports[`Graph Pretty Printing should pretty print with custom options 1`] = `
"MESH_OBJECT
└─ NKNOWN_NODE
  └─ NKNOWN_NODE
    ├─ NKNOWN_NODE
    └─ ASIC_MATERIAL
"
`;

exports[`Graph Pretty Printing should show dependency chain execution 1`] = `
"mesh (node-47) [deps: 1]
└─ unction (node-46) [deps: 1]
  └─ unction (node-45) [deps: 1]
    └─ unction (node-44) [deps: 1]
      └─ unction (node-43) [deps: 1]
        └─ unction (node-42) [deps: 2]
          ├─ onstant (node-40) [deps: 0]
          └─ aterial (node-41) [deps: 0]
"
`;
